Improved TMC2208 stepper motor driver support

LastDragon-ru opened the pull request:
Based on the original #1329 request, improvements:

* Optimized TMC2208 configuration (reduced number of writes)
* Added one-wire write-only mode (so you need only one pin to configure all drivers)
* Added option to disable whole module
  ```
  motor_driver_control.enable                      true
  ```
* Fixed "Alternatively the Smoothieboard must be hard reset after power is on."
  ```
  motor_driver_control.timeout                     0
  ```
* Added SpreadCycle settings
  ```
  motor_driver_control.driver.spreadcycle_toff           5           # TOFF   (2 - 15)
  motor_driver_control.driver.spreadcycle_tbl            32          # TBL    (16/24/32/48)
  motor_driver_control.driver.spreadcycle_hstrt          4           # HSTART (1 - 8)
  motor_driver_control.driver.spreadcycle_hend           0           # HEND   (-3 - 12)
  ```
* Added StealthChop settings
  ```
  motor_driver_control.driver.stealthchop_tpwmthrs        0           # TPWMTHRS
  motor_driver_control.driver.stealthchop_pwm_lim         12          # PWM_LIM
  motor_driver_control.driver.stealthchop_pwm_reg         1           # PWM_REG
  motor_driver_control.driver.stealthchop_freewheel       0           # freewheel
  motor_driver_control.driver.stealthchop_pwm_autograd    1           # PWM_AUTOGRAD
  motor_driver_control.driver.stealthchop_pwm_autoscale   1           # PWM_AUTOSCALE
  motor_driver_control.driver.stealthchop_pwm_freq        1           # PWM_FREQ
  motor_driver_control.driver.stealthchop_pwm_grad        0           # PWM_GRAD
  motor_driver_control.driver.stealthchop_pwm_ofs         36          # PWM_OFS
  ```
* Fixed settings description

tiagojbalmeida committed:
Basic UART communication with TMC2208 stepper motor driver

tiagojbalmeida committed:
Define and document all registers

tiagojbalmeida committed:
Move SoftSerial and Modbus to libs folder

tiagojbalmeida committed:
Use Software UART libs to communicate with TMC2208 driver

tiagojbalmeida committed:
Import functions to configure driver via UART

tiagojbalmeida committed:
Add StealthChop and StealthChop threshold functions

tiagojbalmeida committed:
Add TMC2208 config file

tiagojbalmeida committed:
Update velocity threshold

tiagojbalmeida committed:
Move print information about successfull configured motor drivers

tiagojbalmeida committed:
Add TMC2208 stepper motor driver support

tiagojbalmeida committed:
Move Modbus lib back to spindle module

tiagojbalmeida committed:
Put SPI and UART pins in an union to save memory

tiagojbalmeida committed:
Delete UART object after initial driver configuration

tiagojbalmeida committed:
Put SPI and UART objects inside union to save memory

tiagojbalmeida committed:
Fix SPI/UART union structure

tiagojbalmeida committed:
Remove driver setup condition

tiagojbalmeida committed:
Disable stealthChop threshold as default

tiagojbalmeida committed:
Set general configuration register

tiagojbalmeida committed:
set slaveconf register

tiagojbalmeida committed:
set HOLD parameters

tiagojbalmeida committed:
Add tpowerdown register

tiagojbalmeida committed:
Fix mode parameter name

LastDragon-ru committed:
Added `motor_driver_control.timeout`.

LastDragon-ru committed:
Added `motor_driver_control.driver.tpwmthrs`.

LastDragon-ru committed:
Fixed `motor_driver_control.driver.microsteps` description.

LastDragon-ru committed:
Write only mode support.

LastDragon-ru committed:
Config description and small fixes

LastDragon-ru committed:
Optimized chip configuration (reduced number of writes)

LastDragon-ru committed:
More settings:
* `motor_driver_control.enable`
* `motor_driver_control.driver.tpwmthrs` renamed to `motor_driver_control.driver.stealthchop_tpwmthrs`
* `motor_driver_control.driver.spreadcycle_toff`
* `motor_driver_control.driver.spreadcycle_tbl`
* `motor_driver_control.driver.spreadcycle_hstrt`
* `motor_driver_control.driver.spreadcycle_hend`

LastDragon-ru committed:
Forced `setStepInterpolation(1)`

LastDragon-ru committed:
Stealthchop settings support.

wolfmanjm commented:
This needs to be a patch to the original not a replacement for it. I'm not sure I can just replace the original with this.

LastDragon-ru commented:
@wolfmanjm, as you can see my PR fully based on original, I just added few (required for me) settings (especially tpwmthrs). But

1. I don't know c++ 😃 
2. your notes about deleting pin still actual
3. I'm not sure that writing to chip always works
4. "Alternatively the Smoothieboard must be hard reset after power is on." - makes original PR totally useless (sorry, but this is true), I just added `wait()` and seems it works now...
5. the tpwmthrs setting is required for 2208, the driver useless for normal printing speed without it - (in my case) this the main reason to use drivercontroll.
6. as I understand tpwmthrs also requires spreadcycle configuration, original PR allow configuring only one mode, and you can't set tpwmthrs without additional gcodes. 
7. I see this and not sure that this is good, looks like a conflict (X/Y/etc uses as axis and option?)
    ```
    // MotorDriverControl.cpp
    // M911.1 Pn (or X0)
    // M911.3 S1 Unnn Vnnn Wnnn Xnnn Ynnn setSpreadCycleChopper
    // }else if( (gcode->has_letter('P') && gcode->get_value('P') == id) || gcode->has_letter(axis)) {
 
    // TMC22X.cpp
    // if(s == 0 && HAS('U') && HAS('V') && HAS('W') && HAS('X')) { 
    // s == 4 && HAS('U') && HAS('V') && HAS('W') && HAS('X') && HAS('Y') && HAS('Z')
    ```

So, probably you shouldn't merge any of this or original PR...

fishpepper commented on src/modules/utils/motordrivercontrol/MotorDriverControl.cpp:
```diff
@@ -112,34 +127,86 @@ bool MotorDriverControl::config_module(uint16_t cs)
+
+        if(sw_uart_rx_pin->connected()) {
+            // Read/Write mode
+            PinName txd = port_pin((PortName)(sw_uart_tx_pin->port_number<<8), sw_uart_tx_pin->pin);
```
```suggestion
            PinName txd = port_pin((PortName)(sw_uart_tx_pin->port_number), sw_uart_tx_pin->pin);
```

fishpepper commented on src/modules/utils/motordrivercontrol/MotorDriverControl.cpp:
```diff
@@ -112,34 +127,86 @@ bool MotorDriverControl::config_module(uint16_t cs)
+        if(sw_uart_rx_pin->connected()) {
+            // Read/Write mode
+            PinName txd = port_pin((PortName)(sw_uart_tx_pin->port_number<<8), sw_uart_tx_pin->pin);
+            PinName rxd = port_pin((PortName)(sw_uart_rx_pin->port_number<<8), sw_uart_rx_pin->pin);
```
```suggestion
            PinName rxd = port_pin((PortName)(sw_uart_rx_pin->port_number), sw_uart_rx_pin->pin);
```

fishpepper commented on src/modules/utils/motordrivercontrol/MotorDriverControl.cpp:
```diff
@@ -112,34 +127,86 @@ bool MotorDriverControl::config_module(uint16_t cs)
+            this->serial = new BufferedSoftSerial(txd, rxd);
+        } else {
+            // Write only mode
+            PinName txd = port_pin((PortName)(sw_uart_tx_pin->port_number<<8), sw_uart_tx_pin->pin);
```
```suggestion
            PinName txd = port_pin((PortName)(sw_uart_tx_pin->port_number), sw_uart_tx_pin->pin);
```

As long as Port is 0 your code is fine. I am trying to run it on a MKS SGEN_L where they assigned Port 1 to the rx/tx pins. 

Note to others:  Port1 does not allow interrupts, so the RX functionallity does not work. No problem though, just use NC as port/pin name in the config

fishpepper reviewed (changes requested):
The shift by 8 seems to be a copy and paste mistake (copied from the debug print?). This should not be used here. No problem as long as you use port0. I used port1 and was wondering why my config is ignored...

LastDragon-ru committed:
Removed tx/rx `port_number` shifting.

LastDragon-ru commented:
> The shift by 8 seems to be a copy and paste mistake (copied from the debug print?)

🤷‍♂  Code updated. Thanks. Unfortunately, I can't check/compile it because I migrated to skr 1.3 and marlin...

V1pr commented:
This pull is waiting for testing? If so, I'll try to do it with a 2209 when I've some spare time.

arthurwolf commented:
Yes, please don't hesitate to test.

On Wed, Jan 15, 2020 at 10:38 AM V1pr <notifications@github.com> wrote:

> This pull is waiting for testing? If so, I'll try to do it with a 2209
> when I've some spare time.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/Smoothieware/Smoothieware/pull/1377?email_source=notifications&email_token=AAA2SFIT3NC24LBYAKVHIKLQ53KQDA5CNFSM4G426LO2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEI7VNDI#issuecomment-574576269>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAA2SFPVWFATCFYQVY72NWTQ53KQDANCNFSM4G426LOQ>
> .
>


-- 

勇気と気分

vbolshakov commented:
Not work. Just tried on SKR 1.3 board with 0.0 and 0.1 as TX and RX for X axis.
M911 show status even for not existing TMC2208.
Oscilloscope show serial communication, but no params of TCM2208 changed.
Will try on original Smoothieboard, just need to remove it from CNC.

V1pr commented:
> Not work. Just tried on SKR 1.3 board with 0.0 and 0.1 as TX and RX for X axis.
> M911 show status even for not existing TMC2208.
> Oscilloscope show serial communication, but no params of TCM2208 changed.
> Will try on original Smoothieboard, just need to remove it from CNC.

So far I've found the same, but wanted to test it a bit more. One thing is sure and annoying: it show's the details of an unconnected TMC2208 as you've said.

wolfmanjm commented:
As far as I can tell there is no way to tell if a chip is connected or not, so just like SPI it will just get garbage.

wolfmanjm commented:
It looks like this PR is not really very good, I'm closing it in favor of #1329

vbolshakov commented:
It is possible - it is bi-direction communication.
Using this method it is possible to controll TMC2208 https://learn.watterott.com/silentstepstick/configurator/

vbolshakov commented:
> It looks like this PR is not really very good, I'm closing it in favor of #1329

Also not work for me...

LastDragon-ru commented:
> Also not work for me...

Because both PRs are the same, I just added a few new settings interesting for me. In my case, they also worked unstable so I migrate to Marlin and all fine now. Marlin uses https://github.com/teemuatlut/TMCStepper - the awesome lib that supports a lot of tmc steppers + it is active. So maybe Smoothieware should migrate into it instead of its own outdated code?

vbolshakov commented:
> > Also not work for me...
> 
> Because both PRs are the same, I just added a few new settings interesting for me. In my case, they also worked unstable so I migrate to Marlin and all fine now. Marlin uses https://github.com/teemuatlut/TMCStepper - the awesome lib that supports a lot of tmc steppers + it is active. So maybe Smoothieware should migrate into it instead of its own outdated code?

Your version request interrupt capable pin, "original" not.

V1pr commented:
@vbolshakov just poped into my mind: did you check if the pins of the pololu slot are not grounded by default? (so when you plug in the tmc the UART leg(s) are grounded due to the board layout)

vbolshakov commented:
> @vbolshakov just poped into my mind: did you check if the pins of the pololu slot are not grounded by default? (so when you plug in the tmc the UART leg(s) are grounded due to the board layout)

Yes. I was able to use https://learn.watterott.com/silentstepstick/configurator/ as external tool to check if module work.

V1pr commented:
@vbolshakov  on which axis did you test?
from the driver code: "// TODO hardcoded for X need to select ABC as needed" I was testting with Y...

vbolshakov commented:
@V1pr on X and Y, simultaneously and one by one. 
Check with oscilloscope - "M911", "M911.1 X R0" and "M911.1 X R255" generate UART like signal.

V1pr commented:
Thx. I'll try to split the board tx to see, what does it send really (and connect the split's other end to the ftdi)

vbolshakov commented:
@V1pr I try this, but it looks like software UART baud rate not accurate to be captured...
